crates/proto/src/rr/record_data.rs

```diff
@@ -32,6 +32,15 @@ use crate::serialize::binary::*;
 #[cfg(feature = "dnssec")]
 use super::dnssec::rdata::DNSSECRData;
 
+/// RecordData that is stored in a DNS Record.
+pub trait RecordData: Clone + Sized + PartialEq + Eq + fmt::Display {
+    /// Attempts to convert to this RecordData from the RData type, if it is not the correct type the original is returned
```

Comment:
This feels like it will generate a decent amount of boilerplate. Instead of having these methods, maybe there should be a method `fn map<T: RecordData>(self, impl FnOnce(RData) -> Option<T>) -> Option<TypedRecord<T>>` on the `impl TypedRecord<RData>`? (And an equivalent for references, maybe.)

Reply:
I'll revisit this... now that I have some use cases in place it's a little clearer what is needed.

Reply:
For reference here, @djc, I kept the RecordData trait as well as RData, I added some docs as to why both exist.

There's an alternative here that we could go with, but I hesitate because I think for some of the smaller fixed size record-types it would add unnecessary memory overhead. And that's to use a `Box<dyn RecordData>` in the `Record` type, then we could remove RData entirely. We could go in this direction, but it would mean a heap allocation in all use cases... maybe that's ok? I'm not sure. For now, I would argue we leave both and make a decision later as the code gets more use? Thoughts?

Reply:
Yeah, agreed that we don't want to do `Box<dyn RecordData>` at this level.